src/quantile.js

```diff
@@ -29,20 +29,81 @@ var numericSort = require('./numeric_sort');
  * quantile(data, 0.5); //= 9
  */
 function quantile(sample /*: Array<number> */, p /*: Array<number> | number */) {
-    // Sort a copy of the array. We'll need a sorted array to index
-    // the values in sorted order.
-    var sorted = numericSort(sample);
+    var copy = sample.slice();
 
     if (Array.isArray(p)) {
+        // rearrange elements so that each element corresponding to a requested
+        // quantile is on a place it would be if the array was fully sorted
+        multiQuantileSelect(copy, p);
         // Initialize the result array
         var results = [];
         // For each requested quantile
         for (var i = 0; i < p.length; i++) {
-            results[i] = quantileSorted(sorted, p[i]);
+            results[i] = quantileSorted(copy, p[i]);
         }
         return results;
     } else {
-        return quantileSorted(sorted, p);
+        var idx = quantileIndex(copy.length, p);
+        quantileSelect(copy, idx, 0, copy.length - 1);
+        return quantileSorted(copy, p);
+    }
+}
+
+function quantileSelect(arr, k, left, right) {
+    if (k % 1 === 0) {
+        quickselect(arr, k, left, right);
+    } else {
+        k = Math.floor(k);
+        quickselect(arr, k, left, right);
+        quickselect(arr, k + 1, k + 1, right);
+    }
+}
+
+function multiQuantileSelect(arr, p) {
+    var indices = [0];
+    for (var i = 0; i < p.length; i++) {
+        indices.push(quantileIndex(arr.length, p[i]));
+    }
+    indices.push(arr.length - 1);
+    indices.sort(compare);
+
+    var stack = [0, indices.length - 1];
+
+    while (stack.length) {
+        var r = Math.ceil(stack.pop());
+        var l = Math.floor(stack.pop());
+        if (r - l <= 1) continue;
+
+        var m = Math.floor((l + r) / 2);
+        quantileSelect(arr, indices[m], indices[l], indices[r]);
+
+        stack.push(l, m, m, r);
+    }
+}
+
+function compare(a, b) {
+    return a - b;
+}
+
+function quantileIndex(len /*: number */, p /*: number */)/*:number*/ {
+    var idx = len * p;
+    if (p === 1) {
+        // If p is 1, directly return the last index
+        return len - 1;
+    } else if (p === 0) {
+        // If p is 0, directly return the first index
+        return 0;
+    } else if (idx % 1 !== 0) {
+        // If index is not integer, return the next index in array
+        return Math.ceil(idx) - 1;
+    } else if (len % 2 === 0) {
+        // If the list has even-length, we'll return the middle of two indices
+        // around quantile to indicate that we need an average value of the two
+        return idx - 0.5;
+    } else {
+        // Finally, in the simple case of an integer index
+        // with an odd-length list, return the index
+        return idx;
     }
```

Comment:
@tmcw there's logic duplication between this function and `quantileSorted`, but deduplicating would mean either cutting out into a separate file to be used by both, or exporting it here but introducing a cycled dependency. What should we do?